pkg/utils/predicate.go

```diff
@@ -21,144 +21,303 @@ import (
+	return p.ownersOwnerIsGitTrack(e.Meta.GetOwnerReferences())
+}
+
+// Update returns true if the event object owner is a GitTrack
+func (p OwnersOwnerIsGitTrackPredicate) Update(e event.UpdateEvent) bool {
+	return p.ownersOwnerIsGitTrack(e.MetaNew.GetOwnerReferences())
+}
+
+// Delete returns true if the event object owner is a GitTrack
+func (p OwnersOwnerIsGitTrackPredicate) Delete(e event.DeleteEvent) bool {
+	return p.ownersOwnerIsGitTrack(e.Meta.GetOwnerReferences())
+}
+
+// Generic returns true if the event object owner is a GitTrack
+func (p OwnersOwnerIsGitTrackPredicate) Generic(e event.GenericEvent) bool {
+	return p.ownersOwnerIsGitTrack(e.Meta.GetOwnerReferences())
+}
+
+func (p OwnersOwnerIsGitTrackPredicate) ownersOwnerIsGitTrack(ownerRefs []metav1.OwnerReference) bool {
 	gtoList := &farosv1alpha1.GitTrackObjectList{}
-	err = p.client.List(context.TODO(), gtoList)
+	err := p.client.List(context.TODO(), gtoList)
 	if err != nil {
 		// We can't list GTOs so fail closed and ignore the requests
 		return false
 	}
 
+	// build a set of uids
+	gtoSet := make(map[types.UID]farosv1alpha1.GitTrackObject)
+	for _, item := range gtoList.Items {
+		gtoSet[item.UID] = item
+	}
+
 	for _, ref := range ownerRefs {
-		if ref.Kind == "GitTrackObject" && ref.APIVersion == farosGroupVersion {
-			for _, gto := range gtoList.Items {
-				if ref.UID == gto.UID {
-					return p.ownerInNamespacePredicate.ownerInNamespace(gto.GetOwnerReferences())
-				}
-			}
+		if ref.APIVersion != farosGroupVersion {
+			continue
 		}
-		if ref.Kind == "ClusterGitTrackObject" && ref.APIVersion == farosGroupVersion {
-			for _, cgto := range cgtoList.Items {
-				if ref.UID == cgto.UID {
-					return p.ownerInNamespacePredicate.ownerInNamespace(cgto.GetOwnerReferences())
-				}
+		if ref.Kind != "GitTrackObject" {
+			continue
+		}
+		// the gittrackobject should owned by a gittrack or a clustergittrack.
+		// fetch the gittrackobject, then run through its owners to see if there's
+		// a gittrack in there. We shouldn't see any gittracks from namespaces
+		// that we're not managing since cross namespace ownership is disallowed
+		gto, inset := gtoSet[ref.UID]
+		if !inset {
+			// we have an owner reference to a gittrackobject that isn't in our namespace
+			// be cautious and don't handle this object
+			return false
+		}
+		for _, gtoOwner := range gto.GetOwnerReferences() {
+			if gtoOwner.APIVersion == farosGroupVersion && gtoOwner.Kind == "GitTrack" {
+				return true
 			}
 		}
 	}
 	return false
 }
 
-// NewOwnersOwnerInNamespacePredicate constructs a new OwnersOwnerInNamespacePredicate
-func NewOwnersOwnerInNamespacePredicate(client client.Client) OwnersOwnerInNamespacePredicate {
-	return OwnersOwnerInNamespacePredicate{
-		client:                    client,
-		ownerInNamespacePredicate: NewOwnerInNamespacePredicate(client),
+// OwnersOwnerIsClusterGitTrackPredicate filters events to check the owner of the event
+// object is a ClusterGitTrack
```

Comment:
I think this description is slightly wrong? If it is, the method comments will also be incorrect
```suggestion
// object's owner is a GitTrack
```